In [1]:
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

In [2]:
df = pd.read_csv("../data/raw/AAPL.csv")

df.head()

,Date,Close,High,Low,Open,Volume
0,2009-01-02,2.721686,2.730385,2.554037,2.575630,746015200
1,2009-01-05,2.836553,2.884539,2.780469,2.794266,1181608400
2,2009-01-06,2.789767,2.914229,2.770872,2.877641,1289310400
3,2009-01-07,2.729484,2.774170,2.706990,2.753477,753048800
4,2009-01-08,2.780169,2.793666,2.700393,2.712090,673500800


In [3]:
df["Date"] = pd.to_datetime(df["Date"])

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3774 entries, 0 to 3773
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Date    3774 non-null   datetime64[us]
 1   Close   3774 non-null   float64       
 2   High    3774 non-null   float64       
 3   Low     3774 non-null   float64       
 4   Open    3774 non-null   float64       
 5   Volume  3774 non-null   int64         
dtypes: datetime64[us](1), float64(4), int64(1)
memory usage: 177.0 KB


In [4]:
df["Price_Change"] = df["Close"] - df["Open"]

df[["Date", "Open", "Close", "Price_Change"]].head()

,Date,Open,Close,Price_Change
0,2009-01-02,2.575630,2.721686,0.146056
1,2009-01-05,2.794266,2.836553,0.042287
2,2009-01-06,2.877641,2.789767,-0.087874
3,2009-01-07,2.753477,2.729484,-0.023993
4,2009-01-08,2.712090,2.780169,0.068079


In [5]:
df["Sentiment"] = df["Price_Change"].apply(
    lambda x: 1 if x > 0 else 0
)

df[["Price_Change", "Sentiment"]].head()

,Price_Change,Sentiment
0,0.146056,1
1,0.042287,1
2,-0.087874,0
3,-0.023993,0
4,0.068079,1


In [6]:
X = df[["Open", "High", "Low", "Close", "Volume"]]

y = df["Sentiment"]

print(X.head())

print(y.head())

       Open      High       Low     Close      Volume
0  2.575630  2.730385  2.554037  2.721686   746015200
1  2.794266  2.884539  2.780469  2.836553  1181608400
2  2.877641  2.914229  2.770872  2.789767  1289310400
3  2.753477  2.774170  2.706990  2.729484   753048800
4  2.712090  2.793666  2.700393  2.780169   673500800
0    1
1    1
2    0
3    0
4    1
Name: Sentiment, dtype: int64


In [7]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print("Training data:", X_train.shape)
print("Testing data:", X_test.shape)

Training data: (3019, 5)
Testing data: (755, 5)


In [8]:
model = LogisticRegression()

model.fit(X_train, y_train)

print("Model trained successfully")

Model trained successfully


In [9]:
y_pred = model.predict(X_test)

print(y_pred[:10])

[0 0 0 0 0 0 0 0 0 0]


In [10]:
accuracy = accuracy_score(y_test, y_pred)

print("Accuracy:", accuracy)

Accuracy: 0.4503311258278146


In [11]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.45      1.00      0.62       340
           1       0.00      0.00      0.00       415

    accuracy                           0.45       755
   macro avg       0.23      0.50      0.31       755
weighted avg       0.20      0.45      0.28       755



C:\Users\Kenen\Desktop\news-sentiment-analysis-v2\venv\Lib\site-packages\sklearn\metrics\_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
C:\Users\Kenen\Desktop\news-sentiment-analysis-v2\venv\Lib\site-packages\sklearn\metrics\_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
C:\Users\Kenen\Desktop\news-sentiment-analysis-v2\venv\Lib\site-packages\sklearn\metrics\_classification.py:1833: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(ave

from sklearn.metrics import confusion_matrix
import seaborn as sns

cm = confusion_matrix(y_test, y_pred)

plt.figure(figsize=(6,4))
sns.heatmap(cm, annot=True, fmt="d", cmap="Blues")

plt.title("Confusion Matrix")
plt.xlabel("Predicted")
plt.ylabel("Actual")

plt.show()

# Conclusion

This task implemented a basic machine learning model using Logistic Regression to predict stock sentiment based on AAPL stock market features.

The model used:
- Open
- High
- Low
- Close
- Volume

as input features.

Results showed that the model achieved approximately 45% accuracy, which indicates weak predictive performance. The classification report also revealed that the model struggled to predict positive sentiment values correctly.

Possible reasons include:
- Limited feature engineering
- Imbalanced prediction behavior
- Complexity of stock market movement

Despite the low accuracy, the task successfully demonstrated:
- Data preprocessing
- Feature engineering
- Train/test splitting
- Machine learning model training
- Prediction and evaluation
- Confusion matrix visualization

Future improvements may include:
- Adding technical indicators
- Using real financial news sentiment
- Applying advanced machine learning models
- Balancing the dataset